### ISSUE TITLE
Option to set the text color in legend to be same as the line

### ISSUE THREAD

**Issue by DanHickstein** [matplotlib/matplotlib#10720]
As described in this [SO post](https://stackoverflow.com/questions/13828246/matplotlib-color-coded-text-in-legend-instead-of-a-line), it can look nice when the text in the legend is the same color as the line or marker. I find myself doing this manually nearly every time that I create a legend:

```python
leg = ax.legend()
for line, text in zip(leg.get_lines(), leg.get_texts()):
    text.set_color(line.get_color())
```
The result looks like this: 
![image](https://user-images.githubusercontent.com/1107796/37164735-64a7b786-22b8-11e8-9174-28308be917be.png)

While it's obviously not too difficult to accomplish as-is, I was wondering if there was interest to implement this as another argument to the `legend` function, such that the result could be achieved through something like:
```python
ax.legend(matchtextcolor=True)
```
If this seems like a reasonable feature, I would be happy to try to incorporate it.

**Comment by jklymak** [comment]
I'd be 👍; seems like a reasonable feature to me.

I just proposed to add a kwarg to legend as well so you could look to that for guidance as to whats needed: #10715

The only issue I have is that there are order 20 kwargs for legend right now.  Not sure what the proper pythonic thing to do for huge kwarg lists is, because this one seems a bit out of hand!

**Comment by DanHickstein** [comment]
Thanks @jklymak! 

Yeah, the kwarg list for legend is fairly huge. Is that a problem though? Certainly it makes the docstring somewhat unruly, but in normal usage, users are unlikely to use more than a few of the kwargs at a time. 

I'll proceed with the implementation and see if folks think that the feature is worth the additional kwarg.

**Comment by jklymak** [comment]
Not a problem that should disuade you from making a PR.  Look forward to it!  (well, OK, I'd never do this in a plot, but O can see why some would :wink:)

**Comment by efiring** [comment]
@DanHickstein is there a PR for this now?

**Comment by DanHickstein** [comment]
Sorry, I forgot about this. I will prepare a PR when I get a chance. If someone else would rather do this, please feel free to jump in.

**Comment by timhoffm** [comment]
If we go into this, it might make sense to implement a slightly more general parameter `textcolor`. Apart from 
~~~
ax.legend(textcolor='match')
~~~
This would also allow something like
~~~
ax.legend(textcolor='grey')
~~~

Not sure about the naming `textcolor='match'` is not quite clear. Maybe `textcolor='artist'` or `textcolor='artist-color'`?

**Comment by DanHickstein** [comment]
I like the `textcolor='match'` syntax. Basically, if `'match'` is supplied, then the text color matches the line, as I describe above, otherwise the textcolor is simply set to whatever string is supplied.

Maybe `textcolor='linecolor'` is the most descriptive? Although I suppose it could also be `textcolor='markercolor'`...

**Comment by jklymak** [comment]
I like ‘match’.  Good idea @timhoffm!

**Comment by ImportanceOfBeingErnest** [comment]
The bits on the left side of the legend are called "handles" by matplotlib throughout. Hence 

    textcolor="handle" or textcolor="handles" 

would make most sense, right?

The only obstackle is that this will block "handle" ever being used to name a color. (I would have imagined "handle" being of some glitter-brownish type ;-))

**Comment by jklymak** [comment]
I vote for  `labelcolor`,  just because we use `label` rather than `text` for these annotations, and consistency is good.

However, I think `handle` is incorrect.  The "handles" are the pointers to the artists, not the artists themselves, and referring to them as a "handle" in the user facing API would definitely be confusing.

**Comment by DanHickstein** [comment]
In the [docs for legend](https://matplotlib.org/api/_as_gen/matplotlib.pyplot.legend.html) it seems that labels is used for the list of strings that can be supplied to specify the text. The use of "label" versus "text" is somewhat inconsistent. For example, there is a kwarg `labelspacing`, but there is also `handletextpad`.

If we use `labelcolor`, then I'm not sure if that unambiguously specifies the text color or the text background color. So, I guess I think that `textcolor` is the most intuitive, so that would be my vote. But, `labelcolor` is also fine, if that's what people prefer.

**Comment by jklymak** [comment]
I should have quantified my vote as a very small preference.  Textcolor is by no means unacceptable to me.

**Comment by ImportanceOfBeingErnest** [comment]
As said, handles and labels are the two notions used for legends throught the documentation. I.e.

```
handles, labels = ax.get_legend_handles_labels()
ax.legend(handles=handles, labels=labels)
```

handles are the artists themselves. Handlers are the classes that handle the handles. The code from the original post also would rather be

```
leg = ax.legend()

for handle, label in zip(leg.legendHandles, leg.texts):
    label.set_color(handle.get_color())
```

So "handle" is really the correct notion here. One may argue between `labelcolor` and `textcolor` because `text` is the notion used internally. 

This means either

    labelcolor="handle" or textcolor="handle"

I would argue that since the documentation is what people see from the outside, `labelcolor` is more consistent. `handletextpad` is really the only place which breaks consistency (but I'm not advocating for an API change to `handlelabelpad` here).

The handles can be of very different types. We can safely assume that they are artists and hence have a `get_color`; I'm however not so sure if they necessarily have a `get_markerfacecolor`, such that it's not clear what `labelcolor="markeredge"` would refer to when the handle does not have this property. It should still be doable with some `if` statements and defined fallbacks.

**Comment by efiring** [comment]
I think the use of the name "handle" here is a bit archaic and un-pythonic, possibly an unfortunate Matlab-ism, which, fortunately, we have not adopted elsewhere.  It is also harks back to Unix usage of a "file handle".  In both of those cases the handle is a number used in a lookup table to locate the address of what we would now call an object.

**Comment by ImportanceOfBeingErnest** [comment]
I could well imagine that the "handle" really refers to a handle in the sense of a door-knob. 

![image](https://user-images.githubusercontent.com/23121882/42908353-ff6e3d36-8ae0-11e8-8fec-7849f944cc7a.png)


and I don't see anything wrong with that. It is the name that is used for those graphical items in a legend by convention and unless one wants to change the complete nomenclature of the legend one would of course stick to it to keep it consistent.

**Comment by timhoffm** [comment]
`handle` is the term used in the docs. In the end these are just references to an Artist, which we usually identify with the Artist itself. In that sense it's really not a good name.

I advocate `labelcolor` and @DanHickstein's suggestion for values:  
~~~
labelcolor='linecolor'
labelcolor='markercolor'
labelcolor='markeredgecolor'
labelcolor='red'
~~~

This seems quite clear, also for users who do not know the handles teminology (which you don't need for simple legends).

I don't have the conern that `labelcolor` could be seen as ambiguous. It's common within matplotlib and outside (e.g. css) to name the "main" color just "color". That's clearly the text color for texts.

**Comment by afvincent** [comment]
If one allows things like “markeredgecolor”, does it mean that one should also recognize the short version of such parameters (e.g. “mec”)?

**Comment by ImportanceOfBeingErnest** [comment]
Sure, whoever implements this, can just use every possible property:
```
labelcolor='color'
labelcolor='facecolor'
labelcolor='edgecolor'
labelcolor='linecolor'
labelcolor='markerfacecolor'
labelcolor='markeredgecolor'
labelcolor='markerfacecoloralt'
labelcolor='errorbarcolor'
labelcolor='red'
```

**Comment by StevenLi-DS** [comment]
Btw, this `legendHandles` doesn't get mentioned in the doc at all. Could some pls add it?
https://stackoverflow.com/questions/44490815/matplotlib-match-legend-text-color-with-symbol-in-scatter-plot

**Comment by StevenLi-DS** [comment]
I really like this FR and think maybe it can be even easier for users to modify the text as well as the frame in a legend. No matter how you wanna change the text, the current workflow requires you to loop over the text. For example:

```python
leg = ax.legend()
for text in leg.get_texts:
    text.set_color('red')
    text.set_weight('bold')
    text.set_alpha(0.5)
```

It would be much easier if we can pass those methods as a dict (like a `fontdict`):
```python
textdict = dict(color='red', weight='bold', alpha=0.5)
ax.legend(textdict=textdict)
```
The same logic could also apply to `frame` like, such as:
```python
framedict = (edgecolor='white', alpha=1)
```

**Comment by ca-scribner** [comment]
I was just thinking to myself how do I after the fact make my legend text color match the line color and searching brought me here.  Did this ever get implemented?  I didn't see it anywhere

**Comment by ImportanceOfBeingErnest** [comment]
This never got implemented. It's also not tagged for any milestone for now. If there is interest in implementing it, please do feel free to create a pull request.

**Comment by jklymak** [comment]
Closed by #15898

**Comment by bytelinker** [comment]
Hello,

how to apply this with 3 y-axis and one common legend?
